pkg/duck/sinks.go

```diff
@@ -17,50 +17,77 @@ limitations under the License.
 package duck
 
 import (
-	"context"
 	"fmt"
 
 	corev1 "k8s.io/api/core/v1"
-	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	"k8s.io/client-go/dynamic"
+	"k8s.io/apimachinery/pkg/api/meta"
 
-	duckapis "github.com/knative/pkg/apis"
-	"github.com/knative/pkg/apis/duck"
+	pkgapisduck "github.com/knative/pkg/apis/duck"
 	duckv1alpha1 "github.com/knative/pkg/apis/duck/v1alpha1"
+
+	"github.com/knative/eventing/pkg/reconciler"
+	"github.com/knative/pkg/tracker"
 )
 
-// GetSinkURI retrieves the sink URI from the object referenced by the given
-// ObjectReference.
-func GetSinkURI(ctx context.Context, dynamicClient dynamic.Interface, sink *corev1.ObjectReference, namespace string) (string, error) {
-	if sink == nil {
+// SinkReconciler is a helper for Sources. It triggers
+// reconciliation on creation of the source's sink.
+type SinkReconciler struct {
+	tracker             tracker.Interface
+	sinkInformerFactory pkgapisduck.InformerFactory
+}
+
+// NewSinkReconciler creates and initializes a new SinkReconciler
+func NewSinkReconciler(opt reconciler.Options, callback func(string)) *SinkReconciler {
+	ret := &SinkReconciler{}
+
+	ret.tracker = tracker.New(callback, opt.GetTrackerLease())
+	ret.sinkInformerFactory = &pkgapisduck.CachedInformerFactory{
+		Delegate: &pkgapisduck.EnqueueInformerFactory{
+			Delegate: &pkgapisduck.TypedInformerFactory{
+				Client:       opt.DynamicClientSet,
+				Type:         &duckv1alpha1.AddressableType{},
+				ResyncPeriod: opt.ResyncPeriod,
+				StopChannel:  opt.StopChannel,
+			},
+			EventHandler: reconciler.Handler(ret.tracker.OnChanged),
+		},
+	}
+
+	return ret
+}
+
+// GetSinkURI registers the given object reference with the tracker and if possible,
+// retrieves the sink URI
+func (r *SinkReconciler) GetSinkURI(sinkObjRef *corev1.ObjectReference, source interface{}, sourceName string) (string, error) {
+	if sinkObjRef == nil {
 		return "", fmt.Errorf("sink ref is nil")
 	}
 
-	rc := dynamicClient.Resource(duckapis.KindToResource(sink.GroupVersionKind()))
-	if rc == nil {
-		return "", fmt.Errorf("failed to create dynamic client resource")
+	if err := r.tracker.Track(*sinkObjRef, source); err != nil {
+		return "", fmt.Errorf("Error tracking sink '%+v' for source %q: %+v", sinkObjRef, sourceName, err)
```

Comment:
Is sourceName sufficient? Or do we need name + namespace + type?

Reply:
@Harwayne could you take another look at this area, commit c333e01b19193e70ce523984f9ffb13ad18c7eea?

Reply:
Looks good.

Reply:
Cool! You wouldn't know if there's a function that produces a sourceDesc-like string for any k8s API object in a common format? I looked but I didn't see an obvious candidate.